src/identity_tree.rs

```diff
@@ -1,26 +1,247 @@
+        self.diff.push(TreeUpdate {
+            leaf_index,
+            element,
+        });
+    }
+
+    fn update_without_diff(&mut self, leaf_index: usize, element: Hash) {
+        self.tree.set(leaf_index, element);
+        self.next_leaf = leaf_index + 1;
+    }
+}
+
+#[derive(Clone)]
+pub struct TreeVersion(Arc<RwLock<TreeVersionData>>);
+
+impl From<TreeVersionData> for TreeVersion {
+    fn from(data: TreeVersionData) -> Self {
+        Self(Arc::new(RwLock::new(data)))
+    }
+}
+
+impl TreeVersion {
+    pub async fn peek_next_update(&self) -> Option<TreeUpdate> {
+        let data = self.0.read().await;
+        data.peek_next_update().await
+    }
+
+    pub async fn apply_next_update(&self) {
+        let mut data = self.0.write().await;
+        data.apply_next_update().await;
+    }
+
+    pub async fn update(&self, leaf_index: usize, element: Hash) {
+        let mut data = self.0.write().await;
+        data.update(leaf_index, element);
+    }
+
+    pub async fn next_version(&self) -> Self {
+        let mut data = self.0.write().await;
+        data.next_version()
+    }
+
+    pub async fn append_many_fresh(&self, updates: &[TreeUpdate]) {
+        let mut data = self.0.write().await;
+        let next_leaf = data.next_leaf;
+        updates
+            .iter()
+            .filter(|update| update.leaf_index >= next_leaf)
+            .for_each(|update| {
+                data.update(update.leaf_index, update.element);
+            });
+    }
+
+    pub async fn next_leaf(&self) -> usize {
+        let data = self.0.read().await;
+        data.next_leaf
+    }
+
+    async fn get_proof(&self, leaf: usize) -> (Hash, Proof) {
+        let tree = self.0.read().await;
+        (
+            tree.tree.root(),
+            tree.tree
+                .proof(leaf)
+                .expect("impossible, tree depth mismatch between database and runtime"),
+        )
+    }
+}
+
+pub struct TreeItem {
+    pub status:     Status,
+    pub leaf_index: usize,
+}
+
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Status {
+    Pending,
+    Mined,
+}
+
+#[derive(Debug, Error)]
+#[error("unknown status")]
+pub struct UnknownStatus;
+
+impl FromStr for Status {
+    type Err = UnknownStatus;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "pending" => Ok(Self::Pending),
+            "mined" => Ok(Self::Mined),
+            _ => Err(UnknownStatus),
+        }
+    }
+}
+
+impl From<Status> for &str {
+    fn from(scope: Status) -> Self {
+        match scope {
+            Status::Pending => "pending",
+            Status::Mined => "mined",
+        }
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct InclusionProof {
+    pub status: Status,
+    pub root:   Field,
+    pub proof:  Proof,
+}
+
+#[derive(Clone)]
+pub struct TreeState {
+    mined:  TreeVersion,
+    latest: TreeVersion,
+}
```

Comment:
I would add a todo comment for the additional version explaining why it doesn't exist yet.

Reply:
I thought about it, but in the end I've decided the whole thing should be pretty number-of-versions-agnostic. I can see it being anywhere between 2 and 5 in the coming months.

Reply:
Okay, sounds good!